### ISSUE TITLE
Format devices, entities, and areas as tables for accessibility

### ISSUE THREAD

**Issue by steverep** [home-assistant/frontend#6487]
## Checklist

- [x] I have updated to the latest available Home Assistant version.
- [x] I have cleared the cache of my browser.
- [x] I have tried a different browser to see if it is related to my browser.

## The problem

The configuration pages showing the lists of devices, entities, and areas show the information visibly in tabular format, but they are not semantically marked up as tables in the HTML code.  This is required for assistive technologies such as screen readers to accurately present the content to the user and to allow the user to navigate efficiently.  Not marking them as tables is a failure of the W3C Web Content Accessibility Guidelines (WCAG).  See [WCAG success criterion 1.3.1](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships.html).

## Expected behavior

Tables should be marked up semantically using `<table>`, `<tr>`, and `<td>` (or equivalent ARIA markup).  Relevant WCAG sufficient techniques are:
- [H51: Using table markup to present tabular information](https://www.w3.org/WAI/WCAG21/Techniques/html/H51)
- [H63: Using the scope attribute to associate header cells and data cells in data tables](https://www.w3.org/WAI/WCAG21/Techniques/html/H63)
- [H43: Using id and headers attributes to associate data cells with header cells in data tables](https://www.w3.org/WAI/WCAG21/Techniques/html/H43)

## Steps to reproduce

Try using a screen reader such as the free open source NVDA software to quickly work with and edit devices or entities.  It's painful.

## Environment

- Home Assistant release with the issue: Unknown
- Last working Home Assistant release (if known): Unknown
- Browser and browser version: This problem is browser independent
- Operating system: This problem is OS independent

**Comment by bramkragten** [comment]
We can't use tables because of the virtualization we are applying, but I think setting the correct roles would do the same for a screenreader? `table`, `row`, `cell`?

**Comment by steverep** [comment]
> We can't use tables because of the virtualization we are applying

Do you mean you cannot use the HTML tags `<table>`, `<tr>`, etc., and therefore need to use ARIA?

**Comment by bramkragten** [comment]
Yes

**Comment by steverep** [comment]
Then yes, applying ARIA roles and properties correctly would produce an accessible format as well.  You should either apply the [ARIA table pattern](https://www.w3.org/TR/wai-aria-practices-1.1/#table), or the [ARIA data grid pattern](https://www.w3.org/TR/wai-aria-practices-1.1/#grid).  Which one to choose sort of depends on how many current (or potential) interactive elements the table may have.  If it's zero or otherwise simple, then a table makes sense.  If it's a lot, then a data grid makes sense.

**Comment by steverep** [comment]
Looks as though they are now recognized as tables by NVDA, but the keyboard behavior seems a little wacky.  I or someone else should double check the ARIA implementation.

**Comment by steverep** [comment]
This should not be closed.  My last comment still stands correct.

**Comment by spacegaier** [comment]
You mentioned in your second last comment that you could double check the implementation. Would be great if you could do that and report your findings.

**Comment by steverep** [comment]
Can you point me to the source files which control those tables?

The behavior now is that they are recognized as tables, but navigating them with standard screen reader keyboard shortcuts results in a lot of jumping around ((e.g. using ctrl+alt+↓ should move the focus down one row, but beyond the first several rows it can jump several rows).  I cannot see visually what is happening, but can you tell me if the rows in the table are being loaded on page load or dynamically based on user interaction, such as in a long feed?

**Comment by bramkragten** [comment]
They are in a virtual list, and it has some filters and search, but should by default just show all your devices/entities/areas.

**Comment by steverep** [comment]
> They are in a virtual list, so are only rendered when they would be visible

Not sure what you mean by "virtual list".  Is everything in the DOM on page load?  If not, it would need to be structured as an ARIA feed, but that seems way too complicated for this application (it was meant for article feeds like blogs, RSS, social media, etc.).

**Comment by bramkragten** [comment]
It is not all in the DOM, only the parts that would be on screen are in the DOM.

**Comment by steverep** [comment]
Unless there's a real performance benefit to doing it that way, I would strongly suggest just getting it all in the DOM on page load.  It's not like you're loading significant data with each row as can be the case with feeds.  There's nothing wrong with a big table, and if size is really a concern then just allow the user to paginate it.

**Comment by bramkragten** [comment]
Unfortunately, there is, you can notice it on the states dev tools, that is not virtualized, and is pretty sluggish with a lot of entities in the system.

**Comment by bramkragten** [comment]
I will read into the ARIA feed, but I think the way I implemented it now, with giving the total number of rows and the index of the current row, was the way it was supposed to be done for virtualized tables.

**Comment by steverep** [comment]
No, a feed is not the right ARIA pattern for this.

So I'm starting to look closely at the DOM and attributes now.  The first error I see is that you start `aria-rowindex` at 0 and do not set it on the header row.  Per the spec, `aria-rowindex` should start at 1 and the header row must have an index as well.  So the header row is 1, the first entity/device is 2, and so on.

Similarly, `aria-rowcount` should include the header row.

**Comment by bramkragten** [comment]
Ok thanks for looking into that! Will adjust that.

**Comment by steverep** [comment]
That mistake may very well be the crux of the navigation issue.  [Section 7 of the authoring practices](https://www.w3.org/TR/wai-aria-practices-1.1/#gridAndTableProperties) warns of the "devastating effects" of getting the indexing and count wrong.  I'll test again after it's fixed.

In the meantime, here are some other accessibility improvements to those tables to really make them screen reader friendly:

1. Add the `aria-sort` attribute to the column header of the currently sorted column (e.g. `aria-sort="ascending"` or `aria-sort="descending"`).  It should only be set on one column header at a time.
2. Change the role of the cell that contains the name of the device or entity from `role="cell"` to `role="rowheader"`.  This helps a lot when navigating a table vertically.
3. Save a few bytes of confusion and remove `scope="col"` from the column headers.  It's redundant given the role and is technically only valid on `th` elements.

**Comment by bramkragten** [comment]
I implemented all but the second, as that is a bit harder to do now, the element itself is not aware of the content. But it is something we can add in the future.

**Comment by steverep** [comment]
Please keep this active a bit longer until the 2nd point above can be implemented and I have further time to test the fixes.  Fixing the row index has made a difference, but there is still something off about the keyboard navigation through the table.  I'm trying to figure out the rhyme or reason behind it.

**Comment by steverep** [comment]
This should remain open.  My last comment still applies.

**Comment by steverep** [comment]
I figured out at least part of the remaining cause of navigation issues: empty cells.  These occur in the devices table when a device has no area, and in the entities table when there is no area or status.

Technically, this seems to be a bug in the NVDA screen reader.  I just reported nvaccess/nvda#13110.

However, IMHO these empty cells ought to be filled with something useful and searchable anyway (e.g. "None" when a device or entity has no area).  Something instead of nothing would also be consistent with the battery column for devices, which has a "-" instead of emptiness.

**Comment by Nardol** [comment]
I've just tried to navigate on the devices table using Firefox with Orca (Debian) with no issue.
I did not test other table.
On the Home Assistant side, what other accessibility issues are there with tables?
The bug you mention should be fixed by NVAccess better than on Home Assistant IMHO.

**Comment by Nardol** [comment]
@steverep if "-" were put in empty areas and icon would all have a title, would this issue would be fixed for you?
If yes @bramkragten or @spacegaier could you assign this issue to me?
I'm curently working on it:
* Done for devices, looks like the only potentially empty information is the area when not assigned
* Work in progress for entities, the main issue is to set title for each icon. It is OK for state icon.

I plan to make a PR when all will be done.

**Comment by steverep** [comment]
> if "-" were put in empty areas and icon would all have a title, would this issue would be fixed for you?

No, row headers still need to be marked up.  See 2nd point in [my comment above](https://github.com/home-assistant/frontend/issues/6487#issuecomment-794429582).

**Comment by Nardol** [comment]
So my best won't be to fix this issue but to do what I mentionned above.

**Comment by steverep** [comment]
Nope, not fully fixed yet.

**Comment by Nardol** [comment]
Row headers are missing so as fare as I know, it is not fixed.

**Comment by steverep** [comment]
Still tasks left to complete
